performance_manager/lib/rt_trip_updates.py

```diff
@@ -0,0 +1,283 @@
+    """
+    function to be used with pandas `apply` dataframe method
+    explode nested list of dicts in stop_time_update column
+    """
+    append_dict = {
+        "timestamp": row["timestamp"],
+        "direction_id": row["direction_id"],
+        "route_id": row["route_id"],
+        "start_date": row["start_date"],
+        "start_time": row["start_time"],
+        "vehicle_id": row["vehicle_id"],
+    }
+    return_list: List[Dict[str, Any]] = []
+    for record in row["stop_time_update"]:
+        try:
+            arrival_time = int(record["arrival"]["time"])
+        except (TypeError, KeyError):
+            continue
+        # filter out stop event predictions that are too far into the future
+        # and are unlikel to be used as a final stop event prediction
+        # (2 minutes) or predictions that go into the past (negative values)
+        if (
+            arrival_time - row["timestamp"] < 0
+            or arrival_time - row["timestamp"] > 120
+        ):
+            continue
+        append_dict.update(
+            {
+                "stop_id": record.get("stop_id"),
+                "stop_sequence": record.get("stop_sequence"),
+                "timestamp_start": arrival_time,
+            }
+        )
+        return_list.append(append_dict.copy())
+
+    if len(return_list) == 0:
+        return None
+
+    return return_list
+
+
+def log_memory_usage(complete_queue: SimpleQueue, interval: int = 5) -> None:
+    """
+    log current amount of system memory used every 'interval' seconds
+    to be used in thread to measure work during high memory operation
+    """
+    while True:
+        mb_memory_used = int(psutil.virtual_memory().used / 1_000_000)
+        logging.info("%d MB of memory currently used", mb_memory_used)
+        try:
+            complete_queue.get(timeout=interval)
+        except Empty:
+            continue
+        else:
+            break
+
+
+def unwrap_tu_dataframe(events: pandas.DataFrame) -> pandas.DataFrame:
+    """
+    unwrap and explode trip updates records from parquet files
+    parquet files contain stop_time_update field that is saved as list of dicts
+    stop_time_update must have fields extracted and flattened to create
+    predicted trip update stop events
+    """
+
+    # store start_date as int64
+    events["start_date"] = pandas.to_numeric(events["start_date"]).astype(
+        "int64"
+    )
+
+    # store direction_id as int64
+    events["direction_id"] = pandas.to_numeric(events["direction_id"]).astype(
+        "int64"
+    )
+
+    # store start_time as seconds from start of day int64
+    events["start_time"] = (
+        events["start_time"].apply(start_time_to_seconds).astype("int64")
+    )
+
+    # start thread to log memory usage during apply operation
+    complete_queue: SimpleQueue = SimpleQueue()
+    log_thread = threading.Thread(
+        target=log_memory_usage, args=(complete_queue,)
+    )
+    log_thread.start()
```

Comment:
create and start thread that will be used for logging memory usage during `apply` method